cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,811 @@
+    unique_col_ids.begin(),
+    column_categories.begin(),
+    thrust::equal_to<size_type>(),
+    [] __device__(NodeT type_a, NodeT type_b) -> NodeT {
+      auto is_a_leaf = (type_a == NC_VAL || type_a == NC_STR);
+      auto is_b_leaf = (type_b == NC_VAL || type_b == NC_STR);
+      // (v+v=v, *+*=*,  *+v=*, *+#=E, NESTED+VAL=NESTED)
+      // *+*=*, v+v=v
+      if (type_a == type_b) {
+        return type_a;
+      } else if (is_a_leaf) {
+        // *+v=*, N+V=N
+        // STRUCT/LIST + STR/VAL = STRUCT/LIST, STR/VAL + FN = ERR, STR/VAL + STR = STR
+        return type_b == NC_FN ? NC_ERR : (is_b_leaf ? NC_STR : type_b);
+      } else if (is_b_leaf) {
+        return type_a == NC_FN ? NC_ERR : (is_a_leaf ? NC_STR : type_a);
+      }
+      // *+#=E
+      return NC_ERR;
+    });
+  rmm::device_uvector<TreeDepthT> column_levels(0, stream);  // not required
+  rmm::device_uvector<NodeIndexT> parent_col_ids(num_columns, stream);
+  rmm::device_uvector<SymbolOffsetT> col_range_begin(num_columns, stream);  // Field names
+  rmm::device_uvector<SymbolOffsetT> col_range_end(num_columns, stream);
+  thrust::unique_by_key_copy(
+    rmm::exec_policy(stream),
+    col_ids.begin(),
+    col_ids.end(),
+    thrust::make_zip_iterator(
+      thrust::make_permutation_iterator(tree.parent_node_ids.begin(), node_ids.begin()),
+      thrust::make_permutation_iterator(tree.node_range_begin.begin(), node_ids.begin()),
+      thrust::make_permutation_iterator(tree.node_range_end.begin(), node_ids.begin())),
+    thrust::make_discard_iterator(),
+    thrust::make_zip_iterator(
+      parent_col_ids.begin(), col_range_begin.begin(), col_range_end.begin()));
+  // Restore the order
+  thrust::sort_by_key(rmm::exec_policy(stream),
+                      node_ids.begin(),
+                      node_ids.end(),
+                      thrust::make_zip_iterator(col_ids.begin(), row_offsets.begin()));
+  // convert parent_node_ids to parent_col_ids
+  thrust::transform(rmm::exec_policy(stream),
+                    parent_col_ids.begin(),
+                    parent_col_ids.end(),
+                    parent_col_ids.begin(),
+                    [col_ids = col_ids.begin()] __device__(auto parent_node_id) -> size_type {
+                      return parent_node_id == parent_node_sentinel ? parent_node_sentinel
+                                                                    : col_ids[parent_node_id];
+                    });
+  // copy lists' max_row_offsets to children.
+  // all structs should have same size.
+  thrust::transform_if(
+    rmm::exec_policy(stream),
+    unique_col_ids.begin(),
+    unique_col_ids.end(),
+    max_row_offsets.begin(),
+    [column_categories = column_categories.begin(),
+     parent_col_ids    = parent_col_ids.begin(),
+     max_row_offsets   = max_row_offsets.begin()] __device__(size_type col_id) {
+      auto parent_col_id = parent_col_ids[col_id];
+      while (parent_col_id != parent_node_sentinel and
```

Comment:
If this ever became a performance issue for complex structs, you could consider caching the struct roots with a parallel path compression-like approach.

Reply:
won't it be susceptible to RAW, WAR race conditions? (In this case, it may not matter since the path will end up correct anyway.)
Also, will do it in another PR for optimizations.

Reply:
Yes, you'd probably need atomics to do it safely. But this may not be worth the effort anyways, just a random idea :)